### PR TITLE
Allow Registration of Derivative with Private License Terms

### PR DESCRIPTION
**Pull request by kingster-will**
## Description
This PR introduces changes to the `registerDerivativeIp` function in the `LicenseRegistry` contract. The function now includes a new parameter `isUsingLicenseToken` that allows the registration of a derivative with license terms that are minted by the IP owner but not attached to the IP Asset.

### Changes:

1. Modified the `registerDerivativeIp` function in the `LicenseRegistry` contract to include a new parameter `isUsingLicenseToken`. When `isUsingLicenseToken` is set to true, it allows the registration of a derivative with license terms that are not attached to the IP Asset.

2. Updated the unit tests to reflect these changes. The tests now cover scenarios where an attempt is made to register a derivative with license terms that are not attached to the IP Asset.

## Test Plan 
Add unit tests cover all both positive and negative scenarios of code change in the PR.

## Summary by CodeRabbit

- **New Features**
  - Added support for registering derivative IPs with license tokens in the Licensing and LicenseRegistry modules.

- **Tests**
  - Introduced new tests to verify derivative IP registration with license tokens.

### CODE REVIEW

**Review by Ramarti — APPROVED** [review]
LGTM + Approved by auditors

File: contracts/registries/LicenseRegistry.sol [hunk 1]
```diff
@@ -430,10 +440,14 @@ contract LicenseRegistry is ILicenseRegistry, AccessManagedUpgradeable, UUPSUpgr
         }
         // childIp can only register with default license terms or the license terms attached to the parent IP
         if ($.defaultLicenseTemplate != licenseTemplate || $.defaultLicenseTermsId != licenseTermsId) {
-            if ($.licenseTemplates[parentIpId] != licenseTemplate) {
+            address pLicenseTemplate = $.licenseTemplates[parentIpId];
+            if (
+                (isUsingLicenseToken && pLicenseTemplate != address(0) && pLicenseTemplate != licenseTemplate) ||
```
Comment (jdubpark) [inline contracts/registries/LicenseRegistry.sol:445]: Is `pLicenseTemplate != address(0)` duplicate of `pLicenseTemplate != licenseTemplate`? Or is `licenseTemplate == address(0)` true sometimes so we need to check for `pLicenseTemplate != address(0)`?

Comment (kingster-will) [reply contracts/registries/LicenseRegistry.sol:445]: yes, when parent IP has already attached license (`pLicenseTemplate == address(0)`) then the `licenseTemplate == pLicenseTemplate` should be true.  when parent IP has not yet attach any license (`pLicenseTemplate == address(0)`)  as it is private license, in that case `licenseTemplate != pLicenseTemplate` should be true.

**Review by jdubpark — APPROVED** [review]
